fix: "Show Config" button in error dialog does not work during initialization

traeok opened the pull request:
## Proposed changes

Resolves #3273 

This issue can be reproduced in v3.0.2 if you have a Zowe team configuration file with errors when initializing Zowe Explorer. The prompt with "Error encountered when loading your Zowe config" would appear, but the "Show Config" button did nothing when clicked.

This PR not only fixes the issue, but also adds an integration test to ensure this dialog still works during initialization moving forward.

Finally, this fixes a transient failure encountered with one of the integration tests for updating credentials, so the (Ubuntu, 20.x) stage should not fail now unless there is truly an issue during the test.

## Release Notes

Milestone: 3.0.3

Changelog:

- Fixed an issue during initialization where a broken team configuration file caused the "Show Config" action in the error dialog to stop working. [#3273](https://github.com/zowe/zowe-explorer-vscode/issues/3273)

## Types of changes

<!-- What types of changes does your code introduce to Zowe Explorer? Put an `x` in all boxes that apply. -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

<!-- Put an `x` in all boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer. -->

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [x] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`

**Code coverage**

- [x] There is coverage for the code that I have added
- [x] I have added new test cases and they are passing
- [x] I have manually tested the changes

**Deployment**

- [ ] I have added developer documentation (if applicable)
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here): This may need ported to v2 - if I can reproduce an issue, I will create a port.

adam-wolfe commented:
While this does solve the issue with the Show Config button, after fixing the config and saving changes, trying to add a profile to the trees or refresh the trees results in an error:
![image](https://github.com/user-attachments/assets/8b58d876-6297-47ec-b5c2-68e2d8741972)

traeok commented:
> While this does solve the issue with the Show Config button, after fixing the config and saving changes (and without reloading the VS Code window), trying to add a profile to the trees or refresh the trees results in an error (with nothing added to the logs):

@adam-wolfe This appears to be the behavior since v3.0.0 and is not related to the regression from 3.0.1. That said, I can look into resolving it as part of this PR as well.

traeok commented:
@adam-wolfe I found another place where `getProfileInfo` was being called without handling exceptions, should be working as expected now 👍

adam-wolfe reviewed (dismissed):
Looks good. 

After I fix my config and save changes, only one profile is added to the trees, but if I reload the window, all of my profiles are added to the trees. I'm not sure that this is new behavior though.

zFernand0 commented:
> Looks good.
> 
> After I fix my config and save changes, only one profile is added to the trees, but if I reload the window, all of my profiles are added to the trees. I'm not sure that this is new behavior though.

I ran into the same issue. After fixing the error and saving... the profiles are not loaded. However, it's kind of expected since the watcher was not initialized due to the activation error 😅

zFernand0 reviewed (changes requested):
LGTM! 😋 

Thanks for adding e2e tests 🙏 


Requesting changes just because there is a missing license header in this file:
`packages/zowe-explorer/__tests__/__integration__/bdd/step_definitions/dialogs/ShowConfigErrorDialog.steps.ts`

traeok commented:
Sorry, I tested this again in sandbox and ran into the activation error mentioned above - turns out that I was missing another `null` check in a separate place where `getProfileInfo` is called. 

With this change, fixing the profile and saving causes the default profile to be added to all trees. In addition, it resolves the additional "Activation failed" error and the extension successfully activates even if profiles cannot be loaded.

Thanks for testing and the feedback @adam-wolfe @zFernand0, it's been helpful. I did not see the additional activation error in Dev Host, but was able to reproduce it in the sandbox environment.

---

After looking over the current logic for `ProfilesUtils.getProfileInfo`, I think we should consider refactoring it:
- it does not handle errors unrelated to credential manager initialization and doesn't return anything in those cases (implicitly returns `undefined`)
- it initializes a new `ProfileInfo` instance and reads profiles from disk every time it is called, instead of caching the instance that was first created

For now, I've added a couple safe guards:

- `debug` logger call into the `catch` block of `setupDefaultCredentialManager` so we can get additional context at the time an error occurs in that function
- added typedoc to clarify that the function throws, so that other devs are aware before trying to use it without a try/catch block. 

I chose not to make additional changes to this function for now (such as always returning a `profileInfo` instance), as I see its being used in a few different places - I don't want to alter the behavior in those areas.

zFernand0 commented on packages/zowe-explorer/__tests__/__integration__/bdd/features/dialogs/ShowConfigErrorDialog.feature:
```diff
@@ -0,0 +1,7 @@
+
+Scenario: Initializing Zowe Explorer with a broken profile
+    When a user opens Zowe Explorer
+    Then the Show Config dialog should appear
```
I'm afraid that this step didn't work (as-is).
I had to introduce an error in the config file in order for the test to pass.

Curious if we should add a few steps:
```suggestion
    When a user selects "Edit Team Configuration File" # existing step
    Then it will open the config in the editor #existing step
    When the user introduces and error and saves the configuration file # new step
    Then the Show Config dialog should appear
```

zFernand0 replied:
Thanks for the quick discussion offline 🙏

zFernand0 reviewed (approved):
LGTM! 😋 

Now I do see the default profile loading after saving (same as Adam), which indicates that the watcher still works 🙏 

I don't think it's a big deal to have to reload extensions (or vscode itself), hence why I'm approving 🙏 

I also left a small comment on the bdd tests, but nothing that should be addressed right away.  I'm curious how the CI works though 😅 

re: the refactoring.
Thanks for adding the null checks 🙏 
However, I'm still trying to wrap my head around what the refactoring will entail. 😅 
I could benefit from a quick discussion at some point today 🙏 

Sorry for the long review, but LGTM! 😋

traeok commented:
> I don't think it's a big deal to have to reload extensions (or vscode itself), hence why I'm approving 🙏

Are you encountering a reload during your testing? I didn't add any reloads to the logic, so I just want to make sure its behaving properly for you. The extension will activate even if profiles failed to load, and the watcher is registered during the first activation; so once your config is fixed you should be able to continue using ZE as normal.

> re: the refactoring. Thanks for adding the null checks 🙏 However, I'm still trying to wrap my head around what the refactoring will entail. 😅 I could benefit from a quick discussion at some point today 🙏

Definitely happy to discuss offline 😋

JillieBeanSim reviewed (approved):
LGTM! thanks @traeok
